### PR TITLE
component workflows

### PR DESCRIPTION
**Pull request by sabob**
Here is PR for workflow. I ended up writing a bit more than I intended and covered managed vs unmanaged components as well. Use or ignore as you see fit :)

### CODE REVIEW

File: docs/extend/components.md [hunk 1]
```diff
@@ -109,8 +111,386 @@ const AnotherComponent = Ractive.extend({
...
+// Show B, hide A
+parent.set('showA', false);
+parent.set('showB', true);
+```
+
+With this approach we can dynamically switch between the two components, however we have to declare upfront all the components in the template. We cannot for example replace MyCompA with MyCompC at runtime.
+
+While not as flexible as unmanaged components, the managed approach clearly communicates all the available components through it's template.
+
+With managed components we just need an anchor and can attach any Ractive instance to it:
```
Comment (evs-chris) [inline docs/extend/components.md]: unmanaged, I think

File: docs/extend/components.md [hunk 2]
```diff
@@ -109,8 +111,386 @@ const AnotherComponent = Ractive.extend({
...
+
+We can use the [findComponent()](../api/instance-methods.md#ractivefindcomponent) method to lookup an instance of a child component. For example:
+
+```js
+var parent = new Ractive({
+
+    components: { child: ChildComp },
+
+    validate: function( client ) {
+        // Lookup a ninstance of the child component.
```
Comment (evs-chris) [inline docs/extend/components.md]: a ninstance

File: docs/extend/components.md [hunk 3]
```diff
@@ -109,8 +111,386 @@ const AnotherComponent = Ractive.extend({
...
+
+// Still later
+parent.detachChild( compB );
+parent.attachChild( childC, { target: "child" } );
+```
+
+### Workflow
+Ractive provides a variety of ways that parent and child components can communicate with each other.
+
+Components are great for encapsulating features in their own file and provides a way to reuse compponents in an application, or even across different applications.
```
Comment (evs-chris) [inline docs/extend/components.md]: provide

File: docs/extend/components.md [hunk 4]
```diff
@@ -109,8 +111,386 @@ const AnotherComponent = Ractive.extend({
...
+
+Breaking the dependency between components is great for reusability, but comes at a cost since firing and responding to events is more involved than simply calling methods on the parent.
+
+Let's look at the various ways parent and child components can communicate
+
+#### Data mappings
+
+**Data mapping** is a mechanism that allows us to *connect* pieces of data in the parent to data in the child. When a **data mapping** is specified Ractive creates a [link()](../api/instance-methods.md#ractivelink) between the parent and child data.
+
+Changes on one side will reflect on the other and vice versa.
```
Comment (evs-chris) [inline docs/extend/components.md]: There's nothing wrong with this, I just wanted to note that when two keypaths are linked, they are the _same_ data. In earlier versions of ractive, there were multiple sets of models that tried to sync. Now there's just the source model and a number of reference models that defer to it.

File: docs/extend/components.md [hunk 5]
```diff
@@ -109,8 +111,386 @@ const AnotherComponent = Ractive.extend({
...
+
+The parent component can register as a listener to receive the **personChanged** event.
+
+Let's look at how we can do that.
+
+Here is the child template:
+
+```html
+{{#persons}}
+<input value="name" type="text"/>
```
Comment (evs-chris) [inline docs/extend/components.md]: I think there's a missing binding here

File: docs/extend/components.md [hunk 6]
```diff
@@ -109,8 +111,386 @@ const AnotherComponent = Ractive.extend({
...
+    }
+});
+```
+The child has a button that calls the method *notifyChange* passing along the person object.
+
+Inside the method we [fire()](../api/instance-methods.md#ractivefire) the **personChanged** event to the listeners together with the *parent* object.
+
+When Ractive fires events from components their event names are automatically namespaced with the name they are registered with, meaning Ractive will actually fire the event as *child.personChanged*. In the case of *unmanaged* components, Ractive will use the name of the anchor.
+
+The reason for namespacing is so a parent with multiple components firing events under the same name, say *foo*, know from which child the event was fired from.
```
Comment (evs-chris) [inline docs/extend/components.md]: too much from 😁

File: docs/extend/components.md [hunk 7]
```diff
@@ -109,8 +111,386 @@ const AnotherComponent = Ractive.extend({
...
+
+We make use of [ractive.on()](../api/instance-methods.md#ractiveon) to listener to the event *child.personChanged*. Note the event *personChanged* is namespaced with the component name, *child*.
+
+The event callback function receives a [context](../api/context-object.md) as first argument (ractive always passes in context as a first arg) and the *person* object as second argument.
 
+Next we look at a couple of alternative ways the parent and child components can communicate and pass data. Note that these alternatives will likely decrease reusability of components.
 
+### Isolated
+
+If you aren't interested in reusability of a component or would like to prototype a feature, Ractive provides some convenient features that speeds up the development process.
```
Comment (evs-chris) [inline docs/extend/components.md]: speed

File: docs/extend/components.md [hunk 8]
```diff
@@ -109,8 +111,386 @@ const AnotherComponent = Ractive.extend({
...
+        } else {
+            parent.clientIsInvalid( client );
+        }
+    }
+});
+```
+
+Here we get a reference to the parent components through the [findParent()](../api/instance-methods.md#ractivefindparent) method and informing it whether the client is in a valid state or not by invoking either *clientIsValid()* or *clientIsInvalid()*.
+
+__Note:__ when referencing the parent, we are creating a dependency in that any parent that makes use of the child component will have to provide the *clientIsValid* and *clientIsInvalid* methods, even if that parent has no interest in whether the client is valid or not. If *events* was used instead, the parent could simply ignore those events without Javascript throwing error about missing methods!
```
Comment (evs-chris) [inline docs/extend/components.md]: events were

**Comment by evs-chris** [comment]
Nice!

File: docs/extend/components.md (hunk 5 above)
Comment (sabob) [inline docs/extend/components.md]: Thanks! Because of two-way binding the person object in parent would be updated anyway unless twoway is set to false. Practically though, switching off twoway is painful. But hopefully the idea of firing events is still conveyed. Interestingly, doing homework on other libraries sporting one way data flow they have caveats that binding on 'objects' updates the parent anyway.

File: docs/extend/components.md (hunk 4 above)
Comment (sabob) [inline docs/extend/components.md]: With docs I tend to drag everyone to my level, in other words dumb it down ;-)

So how about this (feel free to change it, and I'll update):
**Data mapping** is a mechanism to *expose* a piece of data in the parent to the child. When a **data mapping** is specified Ractive creates a [link()](../api/instance-methods.md#ractivelink) between the parent and child so that the child can *accesss* the parent data through *keypaths*. Changes on one side will reflect on the other and vice versa.

*Note:* specifying a *data mapping* does not create a *copy* of the data in the child. [Linking()](../api/instance-methods.md#ractivelink) simply informs Ractive that two keypaths refer to the same data.

**Comment by fskreuz** [comment]
I think most of the content could go in https://github.com/ractivejs/ractivejs.github.io/pull/73 since it's more focused on the "how to use" aspect than a "how to write" (which is pretty much the focus of the Extend section - how to write, how to register, how to use as fast as possible). This can be reorganized later though.

The docs have been rewritten to a third-person point of view (like a user manual) instead of a second-person point of view (like a blog). Second-person sounds chatty and tends to become long, fast.

Other than that, good exhaustive documentation of managed vs unmanaged. Could be shorter tho. 👍

**Comment by sabob** [comment]
To my mind the docs should be detailed because where do users get more info after that? The "Getting started" is a good place to get going quickly, while the API is good for references, but docs is really where in depth explanation can take place together with some context as to why you would need a feature eg. components, pros, cons pitfalls etc. But it also depends on the target audience, advanced users or beginners.  Ideally after reading the docs, the user moves from beginner to advanced. Personally I like docs covering a feature in full so I can get a clear picture in my head, hence the inclusion of how components should communicate. Otherwise I have to dig around to understand how things hook up.

**Comment by evs-chris** [comment]
People prefer w3schools over mdn!? I mean, the ES spec is painful to navigate, but mdn is pretty decent. I'm pretty sure w3schools used to kick puppies and duct tape kittens together in the dhtml era.

**Comment by fskreuz** [comment]
I know people who do it for "quick reference" because it comes up first in most searches (without prefixing MSDN/MDN/other sources). 😄

**Comment by sabob** [comment]
Perhaps components belong to both categories? Components seem like a core concept of Ractive.

**Comment by fskreuz** [comment]
Hey @sabob , I think I added your suggestion of managed and unmanaged as part of the workflows PR https://github.com/ractivejs/ractivejs.github.io/pull/73 . Does it look good or am I missing the target? 😄

**Comment by fskreuz** [comment]
Closing in favor of https://github.com/ractivejs/ractivejs.github.io/pull/73 .
